beacon_chain/utils/ssz/src/tree_hash.rs

```diff
@@ -0,0 +1,157 @@
+const CHUNKSIZE: usize = 128;
+const HASHSIZE: usize = 32;
+
+pub trait TreeHash {
+    // Note: it would be nice to have a default trait implementation here
+    // i.e. szz_encode(self) - but rust complains it does not know
+    // the size of 'self'.  Not sure if there's a way around this.
+
+    fn tree_hash(&self) -> Vec<u8>;
+}
+
+// python example:  Note - I'm seeing some inconsistencies
+// between this and the 'Tree Hash' section in the SSZ spec.
+// So, I imagine it will change.
+/* def merkle_hash(lst):
+    # Concatenate list into data
+    if len(lst[0]) != next_power_of_2(len(lst[0])):
+        lst = [extend_to_power_of_2(x) for x in lst]
+    data = b''.join(lst)
+    # Add padding
+    data += b'\x00' * (CHUNKSIZE - (len(data) % CHUNKSIZE or CHUNKSIZE))
+    assert len(data) % CHUNKSIZE == 0
+    # Store length (to compensate for non-bijectiveness of padding)
+    datalen = len(lst).to_bytes(32, 'big')
+    # Convert to chunks
+    chunkz = [data[i:i+CHUNKSIZE] for i in range(0, len(data), CHUNKSIZE)]
+    chunkz = [None] * next_power_of_2(len(chunkz)) + chunkz + [b'\x00' * CHUNKSIZE]
+    for i in range(len(chunkz)//2 - 1, 0, -1):
+        chunkz[i] = hash(chunkz[i*2] + chunkz[i*2+1])
+    return hash(chunkz[1] + datalen) */
+
+/**
```

Comment:
For function comments we have been using the triple-forward-slash `///` style, which works nicely with `rustdoc`. Docs [here](https://doc.rust-lang.org/book/2018-edition/ch14-02-publishing-to-crates-io.html#making-useful-documentation-comments) and local example [here](https://github.com/sigp/lighthouse/blob/77ac5f77bb693aa119952d033a01ebf5ec659c2a/beacon_chain/validation/src/block_validation.rs#L69-L80).